pkg/hive/cell/metric.go

```diff
@@ -0,0 +1,151 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cell
+
+import (
+	"fmt"
+	"reflect"
+	"runtime"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/dig"
+
+	"github.com/cilium/cilium/pkg/hive/internal"
+	pkgmetric "github.com/cilium/cilium/pkg/metrics/metric"
+)
+
+var (
+	withMeta  pkgmetric.WithMetadata
+	collector prometheus.Collector
+)
+
+// Metric constructs a new metric cell.
+//
+// This cell type provides `S` to the hive as returned by `ctor`, it also makes each individual field
+// value available via the `hive-metrics` value group. Infrastructure components such as a registry,
+// inspection tool, or documentation generator can collect all metrics in the hive via this value group.
+//
+// The `ctor` constructor must return a struct or pointer to a struct of type `S`. The returned struct
```

Comment:
Does it make sense to return non-pointer structs at all? I'm guessing that metrics structs will be on the larger side.

Reply:
It depends. The fields of these structs will all be pointers, so its not that big of a deal. Besides, they are mostly transferred during construction, after that they will likely be part of the field of a cell somewhere. I don't think we should restrict it any further if we can help it.